### PR TITLE
8266017: Refactor the *klass::array_klass_impl code to separate the non-exception-throwing API

### PR DESCRIPTION
**Pull request by dholmes-ora**
We currently have array_klass() and array_klass_or_null(), where the former can throw exceptions and the latter can not. But they both delegate to:

array_klass_impl(bool or_null, TRAPS)

which combines both the exception-throwing and non-exception-throwing code into one method declared with TRAPS. To make progress with the change of TRAPS to JavaThread (JDK-8252685) we need to separate these code paths into distinct methods.

Testing: tiers 1-3 (in progress)

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266017](https://bugs.openjdk.java.net/browse/JDK-8266017): Refactor the *klass::array_klass_impl code to separate the non-exception-throwing API

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 80c307c0275ef35feeea973d2f1107031abc0b28
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3711/head:pull/3711` \
`$ git checkout pull/3711`

Update a local copy of the PR: \
`$ git checkout pull/3711` \
`$ git pull https://git.openjdk.java.net/jdk pull/3711/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3711`

View PR using the GUI difftool: \
`$ git pr show -t 3711`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3711.diff">https://git.openjdk.java.net/jdk/pull/3711.diff</a>

</details>

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
This is a nice cleanup. I wish we'd done this years ago!

File: src/hotspot/share/oops/klass.hpp [hunk 1]
```diff
@@ -520,8 +519,10 @@ class Klass : public Metadata {
   inline oop klass_holder() const;
 
  protected:
-  virtual Klass* array_klass_impl(bool or_null, int rank, TRAPS);
-  virtual Klass* array_klass_impl(bool or_null, TRAPS);
+  virtual Klass* array_klass_impl(int rank, TRAPS);
+  virtual Klass* array_klass_impl(TRAPS);
+  virtual Klass* array_klass_or_null_impl(int rank);
+  virtual Klass* array_klass_or_null_impl();
```
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/oops/klass.hpp:525]: I think now we can get rid of the `array_klass_or_null_impl()` function altogether. Instead, `array_klass_or_null` should be declared as `virtual ... = 0` in `Klass` and implemented by the concrete subclasses.

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 2]
```diff
@@ -1448,14 +1446,25 @@ Klass* InstanceKlass::array_klass_impl(bool or_null, int n, TRAPS) {
...
 }
 
-Klass* InstanceKlass::array_klass_impl(bool or_null, TRAPS) {
-  return array_klass_impl(or_null, 1, THREAD);
+Klass* InstanceKlass::array_klass_or_null(int n) {
+  // Need load-acquire for lock-free read
+  if (array_klasses_acquire() == NULL) {
+    return NULL;
+  }
+  // _this will always be set at this point
```
Comment (iklam) [inline src/hotspot/share/oops/instanceKlass.cpp:1457]: I have a hard time understanding the meaning of the above comment. Maybe it should be removed? (Same foe line 1447).

**Review by iklam — APPROVED** [review]
LGTM. Just a small nit about comments.

**Review by coleenp — APPROVED** [review]
Looks great.

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/oops/instanceKlass.cpp:1457]: Until you quoted it I hadn't noticed the underscore on _this. I read it as "this [array_klasses()] will always be set [non-NULL] at this point".

I can restructure this as an if-else to make it crystal clear, and avoid the re-read of array_klasses().

**Review by iklam — APPROVED** [review]
Latest version looks good.

**Pull request closed without merging**